### ISSUE TITLE
"Agent connection server" NullPointerException

### ISSUE THREAD

**Issue by Astarios** [nokia/RED#333]
Hello,

I recently created a project that uses an additional Library to create Dynamic Test Cases. 
Basically at the Runtime of the Tests new ones are created based on an Inputfile.
The Test run normally when I execute them on the commandline, but RED in eclipse shows me an error:

```
An internal error occurred during: "Agent connection server".
java.lang.NullPointerException
```

Also in the console a lot of errors appear that basically have the same error number:
e.g.
```
[ ERROR ] Calling method 'start_keyword' of listener '/tmp/RobotTempDir135732185800854531/TestRunnerAgent.py' failed: error: [Errno 32] Broken pipe
```

Additional Info:
Eclipse version: 2019-06 (4.12.0)
RED version: 0.8.13.201907181006

I created a new project including the library to recreate the problem and attached it as a zip to this report.
Just execute the "Dummy Test Case" in the Error.robot

I don't know if this is an issue in RED itself or if it is a problem with my code and the way it handles the Listener API.

[AgentConnectionError.zip](https://github.com/nokia/RED/files/3707123/AgentConnectionError.zip)

**Comment by michalanglart** [comment]
Hi,

interesting case. Thank you for the project under which we can reproduce the issue. We need to investigate it however at first sight this may be caused by two things (one of them, or even both of them):

1. RED also attaches a listener which uses Listener API v2 called `TestRunnerAgent`. More on it is here: http://nokia.github.io/RED/help/user_guide/launching/red_agent.html and the source code is here: https://github.com/nokia/RED/blob/master/src/RobotFrameworkCore/org.robotframework.ide.core-functions/src/main/python/scripts/TestRunnerAgent.py Maybe both listeners are not working well together.

2. It is possible that there are some assumptions that RED is doing when receiving data from `TestRunnerAgent` and maybe the exception is thrown due to them. The first example that comes to my mind is that for example there is counter in `Execution` view telling what is the progress. Something like `Tests: 123/456`. The last number (number of tests) os of now is stored only when root suite starts and it is never updated. This mean that there may be a mess when you will add tests in the runtime.

We need to investigate the issues with this library.

**Comment by Snooz82** [comment]
@michalanglart 

I saw the same issue.
I am developing a Library that creates TestCases based on CSV or Excel Data files.
robotframework-datadriver (`Library    DataDriver`)

As far as i analyzed the issue:
Listeners are working fine together!
When `Start_Suite` is called, RED-Listener (TestRunnerAgent) sends the Suite information (name and attr) to the REST Server (Eclipse ?!).
in `attr.tests` is a list of all containing TestCases.

I assume that the REST Server remembers the amount of TestCases.
If the numer of the testcase exceeds the precalculated amount of TestCases, the REST Server forces to close the connection of the TestRunnerAgent.

Thats why the TestRunnerAgent from now on raises an exception on every `Start_Suite`

For me this bug has a higher priority and i would be happy to help testing it.

RED is such a awesome Editor for Robot. 

Best Regards
René

**Comment by Snooz82** [comment]
maybe this helps:

RobotFrameworkCore/org.robotframework.ide.core-functions/src/main/java/org/rf/ide/core/execution/agent/event/SuiteStartedEvent.java

The `SuiteStartedEvent` class has a field `childTests` . this contains the TestCases of this suite.

The ExecutionStatusStore (Eclipse-IDE/org.robotframework.ide.eclipse.main.plugin/src/org/robotframework/ide/eclipse/main/plugin/views/execution/ExecutionStatusStore.java) Uses it and calculates the amount of Children.

ExecutionStatusStore.java:137
`    current.addChildren(childSuites.stream()
                .map(childSuite -> new ExecutionTreeNode(current, ElementKind.SUITE, childSuite))
                .collect(toList()));
        current.addChildren(childTests.stream()
                .map(childTest -> new ExecutionTreeNode(current, ElementKind.TEST, childTest, suiteFilePath,
                        childPaths))`


The best way to solve this would be that the Listener of RED would be called as latest.
Problem is, that Command Line Listeners are always the first, and Library Listener are always the last.

**Comment by michalanglart** [comment]
Hi,

It seems that we actually have this issue resolved internally. Currently we are testing it if we can find any issue with fixed implementation. There are many problems with old solution: it assummed that the order of tests or suites will not change, that there may be only tests or only suites on given level, that the number of suites or tests remain unchanged etc. 

This means that there were numerous assumptions which are simply not true and which results in wrongly displayed tree (when order changes) or with exceptions being thrown when number of elements changes. By now it seems all those assumptions in code are removed.

**Comment by Snooz82** [comment]
wow 🥇 awesome!

is someone of you in Robot Framework Slack channel?

**Comment by michalanglart** [comment]
Nope, so far it was sufficient for us to use our issue tracker + RF issue tracker and robotframework-users google group. But - hey, why not :)- I can join RF slack channel. I'll do it tommorrow while in the office and let you know.

If you guys @Astarios @Snooz82 want to test it we can also give you some newest master build so that you could give us a feedback.

BTW. it seems that only `Execution` view was affected by dynamically changing execution model. The other features seem to work. `Message Log` view is quite simple regarding data it processes as it only displays text passed via `log_message` listener method + plus prints which test started/ended, so it's not affected. Additionally debugger could be affected but thankfully it was rewritten almost from scratch around 2 years ago in order to be able to handle remote executions. For remote executions we had to assume that the sources we're having locally can be different than those on different machine which is virtually the same situation as with dynamically changing model.  Thanks to that debugger gracefully handles situations when e.g. a test starts which does not exist in current suite. Of course it will not present source of a test created on-the-fly as it may not exist (although it should handle situaitons when e.g. suite is dynamically added by reading additional .robot file).

**Comment by Astarios** [comment]
Hey, 
thank you for the fast response on my problem. I would be glad to test the new version on my system. Currently I have to switch to a separate console to execute the tests. This would help me very much.

**Comment by michalanglart** [comment]
I've joined RF #general slack channel, so you can reach out to me there.

Here is a not-yet-extensively tested RED with fix for this issue build from current master: 
[RED_feature_0.9.1.201910230724.zip](https://github.com/nokia/RED/files/3761008/RED_feature_0.9.1.201910230724.zip)

**Comment by michalanglart** [comment]
In order to install:
1. Download the feature zip
2. Open `Help -> Install New Software`
3. Click `Add...` button
4. Click `Archive` button
5. Choose downloaded zip file, select new version and proceed with `Next`

**Comment by Snooz82** [comment]
@michalanglart 

This fix works absolutely fine for me.
Debugging and also "normal" Run works without any issues!

No Issues found!

**Comment by michalanglart** [comment]
Glad to hear! Next version will have this change incorporated

**Comment by Astarios** [comment]
> @michalanglart
> 
> This fix works absolutely fine for me.
> Debugging and also "normal" Run works without any issues!
> 
> No Issues found!

Same for me. We have been using the new version for about 5 days now and had no issues. Thank you very much. Great work.

**Comment by ankitaverma1201** [comment]
Hi
@michalanglart Actually I am getting error while implementing Data Driven Testing in Eclipse from excel/csv
I have installed data driver required for fetching data from excel.
But I am getting one error "Test case 'loginwithexcel' contains no keywords to execute" inside *** Test Cases ***.
I have installed data driven library but still getting this error.
@note: **the same code works fine in pycharm**
PFA for the same
![image](https://user-images.githubusercontent.com/57214586/68203208-4fa32480-ffeb-11e9-81f2-5b3fdb509950.png)

**Comment by michalanglart** [comment]
Hi,

> the same code works fine in pycharm

what does it mean? This code works in RED too. You can run it and it should pass. The only difference is that RED validation marks the empty test as error (because normally Robot would fail and tell that empty test case is an error; it doesn't fail here because `DataDriver` modifies this test in the runtime). I don't know that but probably the Robot support for PyCharm doesn't even have something like this (the static code analisys) so it is not suprising that you don't see similar markers.

What you can do:
1) Turn off the validation at all: `Preferences -> Robot Framework -> Errors/Warnings -> Turn off validation`
2) This error is reported as `Robot Framework runtime error` so you can change the severity of the marker: `Preferences -> Robot Framework -> Errors/Warnings -> Runtime and Building -> Runtime -> Robot Framework runtime error` although it would change also severities of other useful problems.

RED cannot statically detect that some empty test will not be empty in runtime. What we could do is to:
1) maybe move this error to separate category (outside of `Robot Framework runtime error`) so that you would change the severity of only this problem
2) implement some mechanism for silencing such problems locally - we do have this idea in the backlog although it's not going to happen very soon probably

**Comment by ankitaverma1201** [comment]
@michalanglart thanks for the solution, although I have find some workaround. Since that test is data driven which it would be fetching from excel during runtime so I initialized the test with empty variables.
PFA for the same.

![image](https://user-images.githubusercontent.com/57214586/68382575-059c7900-017a-11ea-9273-df9cc431e50a.png)

**Comment by Snooz82** [comment]
RED is complaining that without the two arguments at the TestCase it is not valid Robot Syntax.

It still would work, because this test is deleted.

I typically set this test to a default data. A happy path. 
I the datafile is missing, DataDriver will not delete this one and it will be executed. 
It can so be used as the fallback test.

**Comment by michalanglart** [comment]
Fixed in 0.9.1
